### ISSUE TITLE
Default Docker NAT network gateway is missing on Windows Server 2019

### ISSUE THREAD

**Issue by hongyu69** [moby/moby#38358]
**Description**

Docker default nat network gateway is missing after startup, restart the dockerd service can help fix the issue on Windows server 2019. However the issue will come back after reboot. Is there any approach that can make sure dockerd nat network gateway is filled even after reboot? 

**Steps to reproduce the issue:**
1. Install Docker [18.09.0](https://download.docker.com/components/engine/windows-server/18.09/docker-18.09.0.zip) per the steps [https://docs.docker.com/install/windows/docker-ee/#use-a-script-to-install-docker-ee](https://docs.docker.com/install/windows/docker-ee/#use-a-script-to-install-docker-ee)
2. After start dockerd the first time, the subnet of nat network is 0.0.0.0/0, and the gateway is missing
3. Restart dockerd could fix the issue
4. However, this issue will come back again after reboot..

**Describe the results you received:**
PS C:\Users\jesun> docker inspect nat
[
    {
        "Name": "nat",
        "Id": "b7b85981ae2d7f843a9c5fe9535a0bf9f2053dc7de84d825c66e699427b7a275",
        "Created": "2018-12-12T14:20:06.3060877Z",
        "Scope": "local",
        "Driver": "nat",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "windows",
            "Options": null,
            "Config": [
                {
                    "Subnet": "0.0.0.0/0"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {
            "com.docker.network.windowsshim.hnsid": "B145BBB5-14C3-4DE2-9309-73FF64067014",
            "com.docker.network.windowsshim.networkname": "nat"
        },
        "Labels": {}
    }
]

**Describe the results you expected:**
PS C:\Users\jesun> Restart-Service docker
PS C:\Users\jesun> docker inspect nat
[
    {
        "Name": "nat",
        "Id": "1490d59f1065f4aa1986425f25953bd8128a6c159a7c4b0d57f52af724253c5c",
        "Created": "2018-12-12T14:59:28.8669915Z",
        "Scope": "local",
        "Driver": "nat",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "windows",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.22.48.0/20",
                    "Gateway": "172.22.48.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {
            "com.docker.network.windowsshim.hnsid": "B145BBB5-14C3-4DE2-9309-73FF64067014",
            "com.docker.network.windowsshim.networkname": "nat"
        },
        "Labels": {}
    }
]
*The subnet and gateway information here should not be gone after reboot. Ideally the gateway should never be missing.*

**Additional information you deem important (e.g. issue happens only occasionally):**
I ran into the same missing gateway issue on Windows server 2016(10.0.14393.0), and restart dockerd could fix the issue. Most importantly, it seems on Windows server 2016 the fixed gateway and subnet will be persisted, and will not be gone even after reboot. Also Get-NetNat on Windows server 2019 returns nothing even docker nat network is good.

**Output of `docker version`:**

```
PS C:\Users\jesun> docker version
Client:
 Version:           18.09.0
 API version:       1.39
 Go version:        go1.10.3
 Git commit:        33a45cd0a2
 Built:             unknown-buildtime
 OS/Arch:           windows/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.0
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.10.3
  Git commit:       33a45cd0a2
  Built:            11/07/2018 00:24:12
  OS/Arch:          windows/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
PS C:\Users\jesun> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.0
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd gelf json-file local logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 17763 (17763.1.amd64fre.rs5_release.180914-1434)
Operating System: Windows Server 2019 Datacenter Version 1809 (OS Build 17763.134)
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 4GiB
Name: jesun-19
ID: FTGG:C47Y:IXRL:7QZ7:5XS7:HWXO:QSM5:U6DV:BLRN:63VL:XALV:UQEN
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Azure VM and Physical machines

**Comment by hongyu69** [comment]
Just found on Windows server 2019, the newly created hns network will be gone after reboot, which I think is the root cause of the issue I mentioned above.

```
PS C:\Users\jesun\Desktop> New-HnsNetwork -Type NAT -AddressPrefix "192.168.255.0/30" -Gateway "192.168.255.1" -Name "External" -Verbose
VERBOSE: Invoke-HNSRequest Method[POST] Path[/networks] Data[{
    "Subnets":  [
                    {
                        "GatewayAddress":  "192.168.255.1",
                        "AddressPrefix":  "192.168.255.0/30"
                    }
                ],
    "Name":  "External",
    "Type":  "NAT"
}]
VERBOSE: Result :
{"Output":{"ActivityId":"6AB6F657-9F0A-4616-B65A-E00F7B8EFC11","AdditionalParams":{},"CurrentEndpointCount":0,"Extensio
ns":[{"Id":"E7C3B2F0-F3C5-48DF-AF2B-10FED6D72E7A","IsEnabled":false,"Name":"Microsoft Windows Filtering
Platform"},{"Id":"E9B59CFA-2BE1-4B21-828F-B6FBDBDDC017","IsEnabled":false,"Name":"Microsoft Azure VFP Switch
Extension"},{"Id":"EA24CD6C-D17A-4348-9190-09F0D5BE83DD","IsEnabled":true,"Name":"Microsoft NDIS
Capture"}],"Flags":0,"Health":{"AddressNotificationMissedCount":0,"AddressNotificationSequenceNumber":0,"InterfaceNotif
icationMissedCount":0,"InterfaceNotificationSequenceNumber":0,"LastErrorCode":0,"LastUpdateTime":131891524167742904,"Ro
uteNotificationMissedCount":0,"RouteNotificationSequenceNumber":0},"ID":"BDBFFD45-5892-47CA-BD15-2F4551BA93A5","IPv6":f
alse,"LayeredOn":"E71FC4C4-EE06-4CAA-9185-6E5016C2E359","MacPools":[{"EndMacAddress":"00-15-5D-A2-0F-FF","StartMacAddre
ss":"00-15-5D-A2-00-00"}],"MaxConcurrentEndpoints":0,"Name":"External","NatName":"ICS750DF3CE-0580-4E2C-A589-72A9F950A9
AE","Policies":[],"Resources":{"AdditionalParams":{},"AllocationOrder":2,"Allocators":[{"AdapterNetCfgInstanceId":"{750
DF3CE-0580-4E2C-A589-72A9F950A9AE}","AllocationOrder":0,"CompartmendId":0,"Connected":true,"DevicelessNic":false,"Endpo
intNicGuid":"1511D1B4-EB2C-4006-850C-B741C720827A","EndpointPortGuid":"74B0AAF7-F6EE-4662-A8F5-D8351807EA4B","Health":{
"LastErrorCode":0,"LastUpdateTime":131891524170086549},"Hidden":false,"ID":"499AF99C-9866-4B00-96C6-C32C342B0F96","Inte
rfaceGuid":"750DF3CE-0580-4E2C-A589-72A9F950A9AE","IsPolicy":false,"IsolationId":0,"MacAddress":"00-15-5D-A2-04-C8","Ma
nagementPort":true,"NicFriendlyName":"External","PreferredPortFriendlyName":"Container NIC
499af99c","SwitchId":"6A73AFA9-6B50-4ADC-8437-2EC9E2B5DF9E","Tag":"Host
Vnic","WaitForIpv6Interface":false,"nonPersistentPort":false},{"AllocationOrder":1,"Dhcp":false,"Dns":false,"ExternalIn
terfaceConstraint":0,"Health":{"DHCPState":1,"DNSState":1,"ICSState":2,"LastErrorCode":0,"LastUpdateTime":1318915241722
74045},"ICSFlags":0,"ID":"A5E56B72-286C-4CFD-932D-80FF74964E92","IsPolicy":false,"Prefix":30,"PrivateInterfaceGUID":"75
0DF3CE-0580-4E2C-A589-72A9F950A9AE","SubnetIPAddress":"192.168.255.0","Tag":"ICS"}],"Health":{"LastErrorCode":0,"LastUp
dateTime":131891524170086549},"ID":"6AB6F657-9F0A-4616-B65A-E00F7B8EFC11","PortOperationTime":0,"State":1,"SwitchOperat
ionTime":0,"VfpOperationTime":0,"parentId":"8E8AB5F7-17CB-4308-BFDA-87EC8757B195"},"State":1,"Subnets":[{"AdditionalPar
ams":{},"AddressPrefix":"192.168.255.0/30","GatewayAddress":"192.168.255.1","Health":{"LastErrorCode":0,"LastUpdateTime
":131891524167742904},"ID":"A00A3B8D-711E-4673-B44A-B8F5F260C6BA","ObjectType":5,"Policies":[],"State":0}],"TotalEndpoi
nts":0,"Type":"NAT","Version":38654705665},"Success":true}


ActivityId             : 6AB6F657-9F0A-4616-B65A-E00F7B8EFC11
AdditionalParams       :
CurrentEndpointCount   : 0
Extensions             : {@{Id=E7C3B2F0-F3C5-48DF-AF2B-10FED6D72E7A; IsEnabled=False; Name=Microsoft Windows Filtering
                         Platform}, @{Id=E9B59CFA-2BE1-4B21-828F-B6FBDBDDC017; IsEnabled=False; Name=Microsoft Azure
                         VFP Switch Extension}, @{Id=EA24CD6C-D17A-4348-9190-09F0D5BE83DD; IsEnabled=True;
                         Name=Microsoft NDIS Capture}}
Flags                  : 0
Health                 : @{AddressNotificationMissedCount=0; AddressNotificationSequenceNumber=0;
                         InterfaceNotificationMissedCount=0; InterfaceNotificationSequenceNumber=0; LastErrorCode=0;
                         LastUpdateTime=131891524167742904; RouteNotificationMissedCount=0;
                         RouteNotificationSequenceNumber=0}
ID                     : BDBFFD45-5892-47CA-BD15-2F4551BA93A5
IPv6                   : False
LayeredOn              : E71FC4C4-EE06-4CAA-9185-6E5016C2E359
MacPools               : {@{EndMacAddress=00-15-5D-A2-0F-FF; StartMacAddress=00-15-5D-A2-00-00}}
MaxConcurrentEndpoints : 0
Name                   : External
NatName                : ICS750DF3CE-0580-4E2C-A589-72A9F950A9AE
Policies               : {}
Resources              : @{AdditionalParams=; AllocationOrder=2; Allocators=System.Object[]; Health=;
                         ID=6AB6F657-9F0A-4616-B65A-E00F7B8EFC11; PortOperationTime=0; State=1; SwitchOperationTime=0;
                         VfpOperationTime=0; parentId=8E8AB5F7-17CB-4308-BFDA-87EC8757B195}
State                  : 1
Subnets                : {@{AdditionalParams=; AddressPrefix=192.168.255.0/30; GatewayAddress=192.168.255.1; Health=;
                         ID=A00A3B8D-711E-4673-B44A-B8F5F260C6BA; ObjectType=5; Policies=System.Object[]; State=0}}
TotalEndpoints         : 0
Type                   : NAT
Version                : 38654705665



PS C:\Users\jesun\Desktop> Invoke-HNSRequest -Method GET -Type networks


ActivityId             : 6AB6F657-9F0A-4616-B65A-E00F7B8EFC11
AdditionalParams       :
CurrentEndpointCount   : 0
Extensions             : {@{Id=E7C3B2F0-F3C5-48DF-AF2B-10FED6D72E7A; IsEnabled=False; Name=Microsoft Windows Filtering Platform}, @{Id=E9B59CFA-2BE1-4B21-828F-B6FBDBDDC017; IsEnabled=False; Name=Microsoft Azure VFP Switch Extension},
                         @{Id=EA24CD6C-D17A-4348-9190-09F0D5BE83DD; IsEnabled=True; Name=Microsoft NDIS Capture}}
Flags                  : 0
Health                 : @{AddressNotificationMissedCount=0; AddressNotificationSequenceNumber=0; InterfaceNotificationMissedCount=0; InterfaceNotificationSequenceNumber=0; LastErrorCode=0; LastUpdateTime=131891524167742904; RouteNotificationMissedCount=0;
                         RouteNotificationSequenceNumber=0}
ID                     : BDBFFD45-5892-47CA-BD15-2F4551BA93A5
IPv6                   : False
LayeredOn              : E71FC4C4-EE06-4CAA-9185-6E5016C2E359
MacPools               : {@{EndMacAddress=00-15-5D-A2-0F-FF; StartMacAddress=00-15-5D-A2-00-00}}
MaxConcurrentEndpoints : 0
Name                   : External
NatName                : ICS750DF3CE-0580-4E2C-A589-72A9F950A9AE
Policies               : {}
Resources              : @{AdditionalParams=; AllocationOrder=2; Allocators=System.Object[]; Health=; ID=6AB6F657-9F0A-4616-B65A-E00F7B8EFC11; PortOperationTime=0; State=1; SwitchOperationTime=0; VfpOperationTime=0; parentId=8E8AB5F7-17CB-4308-BFDA-87EC8757B195}
State                  : 1
Subnets                : {@{AdditionalParams=; AddressPrefix=192.168.255.0/30; GatewayAddress=192.168.255.1; Health=; ID=A00A3B8D-711E-4673-B44A-B8F5F260C6BA; ObjectType=5; Policies=System.Object[]; State=0}}
TotalEndpoints         : 0
Type                   : NAT
Version                : 38654705665

PS C:\Users\jesun\Desktop> Restart-Computer -Force

PS C:\Users\jesun\Desktop> Invoke-HNSRequest -Method GET -Type networks
```

**Comment by hongyu69** [comment]
Starting from RS5, HNS will not persist NAT and ICS networks any more to prevent leaked nic issues.

**Comment by hongyu69** [comment]
Thanks @olljanat, I think you are right even though I haven't tried. But the key issue here is that nat network gateway field is missing, which is a dependency of one of our partner services that will bind to some port of this gateway so that services running in container can communicate with the service running on local host machine.

**Comment by hongyu69** [comment]
Thanks @olljanat , I got confirmation from Windows HNS team that they will clean up all NAT and ICS internal networks starting from RS5, and the suggestion to fix the issue from them is create the network before starting dockerd. 

I suggested to add back the flag so that NAT and ICS networks can be persisted across reboots, but they denied. Also they said this is a docker team well known zero day issue and current priority of fixing it is low. 

I tried to add a Group Policy startup script to create the network and then start dockerd manually. It works, but is ugly, and who knows someday later the script does not work and dockerd will not get started. 

Now that we cannot restart dockerd as workaround of this gateway missing bug starting from RS5, can you please help prioritize fix of the issue? Thanks a lot.

**Comment by olljanat** [comment]
> can you please help prioritize fix of the issue? Thanks a lot.

@hongyu69 afaik, current situation on Moby project is that:
- there is just one guy from from Microsoft:
https://github.com/moby/moby/blob/46c713ab5ad20b97de01537a8092d2e747f9ad0e/MAINTAINERS#L349-L352
- and one guy from Docker:
https://github.com/moby/moby/blob/46c713ab5ad20b97de01537a8092d2e747f9ad0e/MAINTAINERS#L359-L362

who are able to work with deep technical level Windows related issues like this. Maybe (and hopefully) there is more but I'm not aware of that.

Officially, way to get support / get things prioritised is create premier support case for Microsoft https://success.docker.com/article/where-to-get-help-with-windows but if you have already talk with Windows HNS team then I assume that you have done it already?

How ever I have already said for our premier support contact that Microsoft should connect more people to Moby project.


What comes to me, I'm just the one guy who have been struggling with these Windows containers and Docker related issues soon two years. You can see my very own issue tracker on https://github.com/olljanat/docker-issue-tracking

If we will see this as issue on our environments and it have not been fixed so far then I probably will create PR to fix it but because I have more critical items which we have seen also in production environment on my list I will focus to those first.

PS. Sorry, I know that this was not answer which you was looking for but unfortunately it looks to be situation.

**Comment by olljanat** [comment]
FYI. As there looks to be many issues which prevents users to Win Srv 2019 on production I created EPIC of it to #38498 Please share Microsoft internal tracking when there is any so I can put them to list and people can refer it when contacting support.

**Comment by marknitek** [comment]
We are using a Transparent Network on our Host which gets removed after reboot too.   
To be able to reboot the server if needed, we execute the following Script on Startup (Task Scheduler)
```
$networkName = "LAN"
Write-Output "Check if Network $networkName Exists..."
$ErrorActionPreference = "silentlycontinue"
$inspectResult = & docker network inspect $networkName
$ErrorActionPreference = "continue"
if($LASTEXITCODE -eq 1) {
    Write-Output "Create Container Network $networkName"
    & docker network create -d transparent -o com.docker.network.windowsshim.networkname=Docker-$networkName $networkName
} else {
    Write-Output "Network $networkName already exists"
}

# Get All Containers
$ErrorActionPreference = "silentlycontinue"
$dockerContainers = Invoke-Expression -Command ('docker ps -aq -f "network='+$networkName+'"')
$ErrorActionPreference = "continue"
foreach ($container in $dockerContainers)
{
    # Re-Connect Container to Network
    Write-Output "Connect Container $container to Network $networkName"
    Invoke-Expression -Command "docker network connect $networkName $container"
}
```
Containers get will reconnect to the new network automaticly.

Maybe this helps anyone having similar issues.

**Comment by daschott** [comment]
@marknitek The fix for this is being backported. Can you set the following regkey in the meanwhile?
```
Clean up HNS networks (e.g. docker network rm <id> or get-hnsnetwork | remove-hnsnetwork 
New-ItemProperty HKLM:\SYSTEM\CurrentControlSet\Services\hns\State -Name VolatileNetworks -PropertyType DWORD -Value 1
Restart-service hns
Create new transparent network
```

**Comment by tfenster** [comment]
@daschott I can confirm that the workaround does work for me. Any ideas when the fix you mention will appear? Should it be included if I build master?

**Comment by daschott** [comment]
I believe there is 2 issues conflated here in this thread that show very similar symptoms. One issue is that network ID isn't being updated after reboot on WS2019 which has a change in it so that NAT networks aren't being persisted. This should be fixed via Docker EE 18.09.3 release: https://github.com/docker/engine/pull/149

The other issue is that transparent networks are not being persisted after reboot, which is a Windows platform bug. The RegKey I posted earlier is a workaround for this. I am still confirming which KB will include that fix but will post it here once I know for sure.

**Comment by jhowardmsft** [comment]
@olljanat Can you update your "epic" with the information from @daschott above, and split it into two separate issues along with current status.

**Comment by hongyu69** [comment]
> I believe there is 2 issues conflated here in this thread that show very similar symptoms. **One issue is that network ID isn't being updated after reboot on WS2019 which has a change in it so that NAT networks aren't being persisted. This should be fixed via Docker EE 18.09.3 release: [docker#149](https://github.com/docker/engine/pull/149)**
> 
> The other issue is that transparent networks are not being persisted after reboot, which is a Windows platform bug. The RegKey I posted earlier is a workaround for this. I am still confirming which KB will include that fix but will post it here once I know for sure.

@daschott I tried docker EE 18.09.3 on Windows Server 2019 and the issue is still there. Can you help double confirm the fix?

Thanks.

**Comment by daschott** [comment]
The platform fix for this has been released now: https://support.microsoft.com/en-us/help/4482887/windows-10-update-kb4482887

The platform fix resolves the issue of transparent networks not being persisted after reboot. The Docker EE 18.09.3 is needed to resolve the issue of Docker not referencing the correct network ID after reboot causing containers to fail to re-attach. 

@hongyu69 There should be no more need for the workaround. Can you please install the KB & update to Docker EE 18.09.3 and confirm if you continue seeing issues?

**Comment by BrianThorChristensen** [comment]
Can we get this fix for Docker Desktop for Windows as well?

**Comment by noraab** [comment]
Hi @daschott ,
just upgraded Docker EE to 18.09.3 on Windows Server 2019 and (manually) installed KB4482887 from http://www.catalog.update.microsoft.com/Search.aspx?q=KB4482887.

* Rebooted
* Started some containers using docker-compose
* Rebooted
* Containers not running anymore
* Windows Event Log shows:

`Failed to start container e1e9f87f51682ae206e25620b937610490b7e402203021c133b60efc5ed06606: network 5bd37faf2e56041dc5ffc16422e753438da3f909c4f24c57a99caf7f441e78d9 not found`

From my understanding, networks are still not persisted after reboot...

```
PS C:\Windows\system32> docker --version
Docker version 18.09.3, build 142dfcedca
```
```
PS C:\Windows\system32> docker-compose --version
docker-compose version 1.23.2, build 1110ad01
```
```
PS C:\Windows\system32> Get-ComputerInfo
OsName                                                  : Microsoft Windows Server 2019 Standard
OsType                                                  : WINNT
OsOperatingSystemSKU                                    : StandardServerEdition
OsVersion                                               : 10.0.17763
```

**Comment by wtigotw** [comment]
I have also installed the KB and am running Engine v18.09.3 on Windows 10 build 1809. The problems continue to persist for me as well. After a restart, manually created networks are deleted and need to be re-added; networks related to docker-compose setups exist but their IDs have changed and the containers don't recognize them.

**Comment by hongyu69** [comment]
> Hi @daschott ,
> just upgraded Docker EE to 18.09.3 on Windows Server 2019 and (manually) installed KB4482887 from http://www.catalog.update.microsoft.com/Search.aspx?q=KB4482887.
> 
> * Rebooted
> * Started some containers using docker-compose
> * Rebooted
> * Containers not running anymore
> * Windows Event Log shows:
> 
> `Failed to start container e1e9f87f51682ae206e25620b937610490b7e402203021c133b60efc5ed06606: network 5bd37faf2e56041dc5ffc16422e753438da3f909c4f24c57a99caf7f441e78d9 not found`
> 
> From my understanding, networks are still not persisted after reboot...
> 
> ```
> PS C:\Windows\system32> docker --version
> Docker version 18.09.3, build 142dfcedca
> ```
> ```
> PS C:\Windows\system32> docker-compose --version
> docker-compose version 1.23.2, build 1110ad01
> ```
> ```
> PS C:\Windows\system32> Get-ComputerInfo
> OsName                                                  : Microsoft Windows Server 2019 Standard
> OsType                                                  : WINNT
> OsOperatingSystemSKU                                    : StandardServerEdition
> OsVersion                                               : 10.0.17763
> ```
> @hongyu69, is the problem resolved for you? Could you kindly post some info about your environment?

I am sure NAT and ICS networks are not persisted across reboots starting from RS5 any more, which is by design. I don't find fixes submitted in docker for this empty nat gateway issue, even in RS1.

**Comment by daschott** [comment]
For the users that are seeing issues with persistence of transparent networks, when applying the KB for persistence of transparent networks, please ensure you have:
1. Cleaned up all currently created HNS network
2. Applied KB and rebooted
3. Created a new HNS network

Such that the new code can be applied to create the (new) network correctly. We are investigating other scenarios which may cause the fix to appear as if it hasn't applied.

@hongyu69 Regarding the Docker NAT network gateway field not being populated, this is a Docker issue that is still active. Please don't close this issue.

**Comment by noraab** [comment]
Hey @daschott,

thanks for your last reply. After reading, I'm not a 100% sure if I tested correctly.

- I removed any existing docker networks via `docker network rm`
- Applied KB and rebooted
- New network is created when I call `docker-compose up`

Could you kindly give me a hint regarding how to clean up all HNS networks? Which command do I have to execute?

**Comment by daschott** [comment]
hi @noraab, `docker network rm` should work just fine.  Another way is to call 
`get-hnsnetwork | remove-hnsnetwork` 

If there is a particular network you want to remove:
`get-hnsnetwork | ? Name -Like "my_transparent" | remove-hnsnetwork

I suggest you first try `docker network rm` , then verify using `docker network ls` and `get-hnsnetwork` that the network is gone. If it isn't, you can use `remove-hnsnetwork` to remove it. 

Can you confirm whether you are still seeing transparent networks being persisted after that?

**Comment by noraab** [comment]
Thanks for your reply @daschott,

the steps you provided helped to get rid of any existing networks. Thats fine.
But unfortunatly, containers started using compose still don't come up again after a system reboot.

The steps I executed:
* I removed any existing docker networks via `docker network rm`
* Executed `get-hnsnetwork | remove-hnsnetwork`
* Executed `get-hnsnetwork` again - no networks present anymore (as expected)
* Reboot
* Run `docker-compose up`
* Reboot
* Run `docker-compose ps` - no containers running. Even though the compose file defines `restart: always` for any of the services.

Still in the Windows event log I see:
`Failed to start container fdf7dec28684166d27544660c99296ae3b609a38704755bd41290dce2862a0a8: network 0c0cb08f1710fb961d6949eeb3da4db5f99cb2e85a5d26d1e6d7d0146d77af9c not found`

**Comment by cukal** [comment]
I'm having the same issue on Windows 2019. The default nat network id changes after a reboot and eventviewer throws:

```Failed to start container 3332e53cdcc0ef87d147c60526e7da8126d4dfcd4f3328cc17ddffea32214f31: network c5821f52bbb0378fd9a22db9dd670b3e9c90c34e6e0fb09c0f40479bdae1c283 not found```


```Client: Docker Engine - Enterprise
 Version:           18.09.4
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        c3516c43ef
 Built:             03/27/2019 18:22:15
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Enterprise
 Engine:
  Version:          18.09.4
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.10.8
  Git commit:       c3516c43ef
  Built:            03/27/2019 18:20:29
  OS/Arch:          windows/amd64
  Experimental:     true
```
```
Kernel Version: 10.0 17763 (17763.1.amd64fre.rs5_release.180914-1434)
```
Is there a workaround?

**Comment by noraab** [comment]
Hi @cukal ,
> Is there a workaround?

Currently the only workaround I'm aware of is to use Windows Task Scheduler to run `docker-compose down && docker-compose up -d` on system reboot.

**Comment by olljanat** [comment]
> I believe there is 2 issues conflated here in this thread that show very similar symptoms. One issue is that network ID isn't being updated after reboot on WS2019 which has a change in it so that NAT networks aren't being persisted. This should be fixed via Docker EE 18.09.3 release: [docker#149](https://github.com/docker/engine/pull/149)
> 
> The other issue is that transparent networks are not being persisted after reboot, which is a Windows platform bug. The RegKey I posted earlier is a workaround for this. I am still confirming which KB will include that fix but will post it here once I know for sure.

@daschott looks that those issues are not still fixed. On #39098 user have latest Docker version and latest Win 2019 build but issue still exist (or is that another issue with very similar symptoms?)

cc @akunaatrium

**Comment by olljanat** [comment]
> @marknitek The fix for this is being backported. Can you set the following regkey in the meanwhile?

@daschott what is status with backport?


> I'm having the same issue on Windows 2019. The default nat network id changes after a reboot and eventviewer throws:
> 
> `Failed to start container 3332e53cdcc0ef87d147c60526e7da8126d4dfcd4f3328cc17ddffea32214f31: network c5821f52bbb0378fd9a22db9dd670b3e9c90c34e6e0fb09c0f40479bdae1c283 not found`

@cukal that is actually another issue. Please follow #39098 about it.

**Comment by cukal** [comment]
@olljanat Not sure if I remember correctly but that issue talks about a _custom_ network. I am referencing the _default_ network. It's still present but with another id and that breaks things.

**Comment by olljanat** [comment]
@cukal good point then it is related to this one but I was not able reproduce it on fully patched Win 2019 (win and docker patches) so maybe it is fixed already?

**Comment by NKnusperer** [comment]
Same problem here 😠
This is very annoying. After a reboot, services deployed via a stack have no working network connection because of this, We have to restart Docker like a madman until it randomly starts working again until the next time. This is madness. Please investigate.

**Comment by daschott** [comment]
@hongyu69 for issue that you reported, the fix that is being worked on can be tracked in #2429. 

For users that have scripts that rely on output from `docker network inspect nat`, a potential workaround could also be to check NAT network IP range and gateway using ipconfig or following command:
`netsh interface ip show addresses "vEthernet (nat)"`

**Comment by hongyu69** [comment]
I tried Docker EE 19.03.5 on RS1 and RS5, it seems this is already fixed, thanks for resolving this issue.

**Comment by Masterwow3** [comment]
Hi everyone,

I was able to solve the issue by restoring the HNS networks after a server reboot. This prevents Docker from losing its network configurations on restart, allowing containers to start as expected without any manual intervention.

As a result, any previous network-related workarounds are no longer necessary.

I've made this solution available as a Docker plugin, which can also automatically register itself with the Windows Task Scheduler to ensure the networks are restored automatically after every Windows boot.

You can find the plugin here: https://github.com/Masterwow3/docker-netrestore

**Comment by olljanat** [comment]
@Masterwow3 nice one but if you was able to catch where it goes wrong and how to fix, why not create pull request to Moby and fix issue for everyone?
